symbolic-base/test/Tests/Symbolic/Data/Common.hs

```diff
@@ -0,0 +1,234 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeOperators       #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Tests.Symbolic.Data.Common
+  ( specConstantRoundtrip
+  , specSymbolicFunction0
+  , specSymbolicFunction1
+  , specSymbolicFunction1WithPar
+  , specSymbolicFunction2
+  ) where
+
+import           Data.Binary                      (Binary)
+import           Data.Eq                          (Eq)
+import           Data.Function                    (const, id, ($))
+import           Data.Functor.Rep                 (Representable (..))
+import           Data.Typeable                    (Proxy (..))
+import           GHC.Generics                     (U1 (..), type (:*:) (..))
+import           Prelude                          (String, return, type (~), (++))
+import           Test.Hspec                       (Spec, describe)
+import           Test.QuickCheck                  (Arbitrary (..), Gen, (===))
+import           Tests.Symbolic.ArithmeticCircuit (it)
+import           Text.Show                        (Show)
+
+import           ZkFold.Base.Algebra.Basic.Class  (FromConstant (..), ToConstant (..))
+import           ZkFold.Symbolic.Class            (Arithmetic, Symbolic (BaseField))
+import           ZkFold.Symbolic.Compiler         (ArithmeticCircuit, checkCircuit, checkClosedCircuit, compileWith,
+                                                   eval, exec)
+import           ZkFold.Symbolic.Data.Class       (SymbolicData (..), SymbolicOutput)
+import           ZkFold.Symbolic.Data.Input       (SymbolicInput)
+import           ZkFold.Symbolic.Interpreter      (Interpreter (..))
+
+{-
+  For all symbolic types we need to do the following:
+  1. Check that `fromConstant` and `toConstant` are inverse functions.
+  2. For all related functions, check that the circuit evaluation is equivalent to the function evaluation and
+     that the circuit's constraints are satisfied by the respective witness.
+  3. Run all type-specific checks in the interpreter context.
+-}
+
+specConstantRoundtrip :: forall a x .
```

Comment:
Since `x` is always used here with `Interpreter a`, the type can be generalized a bit:
```haskell
specConstantRoundTrip :: forall x .
    ( Arbitrary x, Eq x, Eq (Const x), Show x, Show (Const x)
    , FromConstant (Const x) x, ToConstant x
    ) => String -> String -> Gen (Const x) -> Spec
```

Reply:
It then suffices to define
```haskell
specConstantRoundtrip = specConstantRoundtrip'
```